lib/ffi-libarchive/reader.rb

```diff
@@ -171,5 +185,57 @@ def save_data(file_name)
         raise Error, @archive if C.archive_read_data_into_fd(archive, fd) != C::OK
       end
     end
+
+    private
+
+    #
+    # See:
+    #   1. https://github.com/libarchive/libarchive/blob/6a9dcf9fc429e2dc9fb08e669bf7b0bed4d5edf9/tar/read.c#L346
+    #   2. https://github.com/libarchive/libarchive/blob/a11f15860ae39ecdc8173243a211cdafc8ac893c/tar/util.c#L523-L535
+    #   3. https://github.com/libarchive/libarchive/blob/a11f15860ae39ecdc8173243a211cdafc8ac893c/tar/util.c#L554-L560
+    #
+    # @param entry [Archive::Entry]
+    #
+    # @return [Archive::Entry, nil] entry stripped or nil if entry is no longer relevant due to stripping
+    #
+    def strip_entry_components(entry)
+      if strip_components > 0
+        name = entry.pathname
+        original_name = name.dup
+        hardlink_name = entry.hardlink
+        original_hardlink_name = hardlink_name.dup
+
+        strip_path_components!(name)
+        return if name.empty?
+
+        unless hardlink_name.nil?
+          strip_path_components!(hardlink_name)
+          return if hardlink_name.empty?
+        end
+
+        if name != original_name
+          entry.copy_pathname(name)
+        end
+        entry.copy_hardlink(hardlink_name) if hardlink_name != original_hardlink_name
+      end
+
+      entry
+    end
+
+    #
+    # @param path [String]
+    #
+    # @return [String]
+    #
+    def strip_path_components!(path)
+      if strip_components > 0
+        is_dir = path.end_with?("/")
+        updated_path = path.split("/").drop(strip_components).join("/")
+        updated_path = is_dir && updated_path != "" ? updated_path + "/" : updated_path
```

Comment:
do we have to worry about backwhacks in zip files on windows?

Reply:
I don't know. I'll test it out and see what these look like on Windows.

Reply:
I just checked and both Windows and Linux uses forward slashes (`/`)